### ISSUE TITLE
Conv2d triggers assertion in mkl-dnn when padding=(n, 3)

### ISSUE THREAD

**Issue by arp2600** [pytorch/pytorch#20583]
## 🐛 Bug

An assertion is triggered:
```
Assertion failed: (jcp.ur_w * (jcp.nb_oc_blocking + 1) <= num_avail_regs), function init_conf, file ../third_party/ideep/mkl-dnn/src/cpu/jit_avx2_conv_kernel_f32.cpp, line 567.
Abort trap: 6
```
When doing:
```
conv = torch.nn.Conv2d(nChans, nChans, kernel_size=kernelSize, padding=(n, 3))
x = conv(inputData)
```

Error occurs when `padding[1] == 3`. The value of `padding[0]` does not seem to matter.

Error occurs on this processor `Intel(R) Xeon(R) CPU E5-2697 v2 @ 2.70GHz`
But not this processor `Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz`

Error occurs using torch version `1.0.1` and `1.1.0`
But not using torch version `1.0.0`

## To Reproduce

1. Save and run the script below.
Script:
```
import torch

nChans = 4
kernelSize = 3
samplesPerBatch = 6
inputShape = [samplesPerBatch, nChans, 32, 32]

inputData = torch.ones(*inputShape, dtype=torch.float32)


def run_conv(padding):
    conv = torch.nn.Conv2d(
        nChans, nChans, kernel_size=kernelSize, padding=padding)
    x = conv(inputData)


# These are all ok
print('Trying conv with padding (1, n)'
      ' where n is all values from 0 to 99 excluding 3')
for i in range(100):
    if i != 3:
        run_conv((1, i))

# This is ok
print('Trying conv with padding (3, 1)')
run_conv((3, 1))

# This one causes an assertion
print('Trying conv with padding (1, 3)')
run_conv((1, 3))
```
Output:
```
$ python ../poponnx/torch_test/model.py
Trying conv with padding (1, n) where n is all values from 0 to 99 excluding 3
Trying conv with padding (3, 1)
Trying conv with padding (1, 3)
Assertion failed: (jcp.ur_w * (jcp.nb_oc_blocking + 1) <= num_avail_regs), function init_conf, file ../third_party/ideep/mkl-dnn/src/cpu/jit_avx2_conv_kernel_f32.cpp, line 567.
Abort trap: 6
```

## Environment
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.6
GCC version: Could not collect
CMake version: version 3.12.4

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.3
[pip3] torch==1.1.0
[pip3] torchvision==0.2.2.post3
[conda] Could not collect

## Additional context

More info on machine where error occurs:
- Model Name: Mac Pro
- Model Identifier: MacPro6,1
- Processor Name: 12-Core Intel Xeon E5
- Processor Speed: 2.7 GHz
- Number of Processors: 1
- Total Number of Cores: 12

**Comment by soumith** [comment]
cc: @mingfeima can someone on intel side reproduce the issue, and if its mkl-dnn bug, issue a fix

**Comment by mingfeima** [comment]
@soumith will try to reproduce on our side.

**Comment by mingfeima** [comment]
@arp2600 Hi, i double checked your script on Xeon E5 2699 (broadwell, avx2) and Xeon 6148 (skylake, avx512), i can not reproduce this issue. The PyTorch version is `1.1.0a0`.

could you please specify how you installed your PyTorch environment? Build from source or conda?

**Comment by arp2600** [comment]
Hi @mingfeima. I installed PyTorch using pip. Below I've included the `pip install` and `pip show` outputs. I notice the two machines you tried to reproduce the issue on both support avx2. The machine I am getting the error on only has support for avx, not avx2.
```
$ pip install torch
Collecting torch
  Using cached https://files.pythonhosted.org/packages/c9/e2/40f7f017437139e5dc076ea5ad4207da61250cc078bceef7d6f333c1d05c/torch-1.1.0-cp37-none-macosx_10_7_x86_64.whl
Requirement already satisfied: numpy in ./lib/python3.7/site-packages (from torch) (1.16.3)
Installing collected packages: torch
Successfully installed torch-1.1.0
$
$ pip show torch
Name: torch
Version: 1.1.0
Summary: Tensors and Dynamic neural networks in Python with strong GPU acceleration
Home-page: UNKNOWN
Author: UNKNOWN
Author-email: UNKNOWN
License: UNKNOWN
Location: /Users/georger/Development/torch_test/python_env/lib/python3.7/site-packages
Requires: numpy
Required-by: torchvision
```

**Comment by huangbiubiu** [comment]
@mingfeima Same problem.
Environment Info:
```
CPU: Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz
PyTorch: 1.1.0
```

**Comment by vpirogov** [comment]
This looks very much like an issue fixed in [Intel MKL-DNN v0.21.1](https://github.com/intel/mkl-dnn/releases/tag/v0.21.1). Observed behavior reproduces in **debug** builds only on Intel AVX systems. Neither Intel AVX2 nor Intel AVX512 are impacted.

**Comment by mingfeima** [comment]
> This looks very much like an issue fixed in [Intel MKL-DNN v0.21.1](https://github.com/intel/mkl-dnn/releases/tag/v0.21.1). Observed behavior reproduces in **debug** builds only on Intel AVX systems. Neither Intel AVX2 nor Intel AVX512 are impacted.

yes, this one should have been fixed with last ideep upgrade. @arp2600 please double check with the latest pytorch code base, let me know if you still have this issue.

**Comment by sitsofe** [comment]
@mingfeima I'm afraid this still looks broken to me:
```
$ sysctl hw.model
hw.model: MacPro6,1
$ python3 -m venv /tmp/torchenv
$ . /tmp/torchenv/bin/activate
$ pip install numpy
Collecting numpy
  Using cached https://files.pythonhosted.org/packages/b4/e8/5ececadd9cc220bb783b4ce6ffaa9266925d37ed41237bc23bc530ab4f3d/numpy-1.17.2-cp37-cp37m-macosx_10_6_intel.whl
Installing collected packages: numpy
Successfully installed numpy-1.17.2
$ pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
Looking in links: https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
Collecting torch
  Using cached https://download.pytorch.org/whl/nightly/cpu/torch-1.3.0.dev20190925-cp37-none-macosx_10_9_x86_64.whl
Installing collected packages: torch
Successfully installed torch-1.3.0.dev20190925
$ python3 -c "import torch; print(torch.__version__)"
1.3.0.dev20190925
$ python3 test.py 
Trying conv with padding (1, n) where n is all values from 0 to 99 excluding 3
Trying conv with padding (3, 1)
Trying conv with padding (1, 3)
Assertion failed: (jcp.ur_w * (jcp.nb_oc_blocking + 1) <= num_avail_regs), function init_conf, file ../third_party/ideep/mkl-dnn/src/cpu/jit_avx2_conv_kernel_f32.cpp, line 567.
Abort trap: 6
```

So I don't think this issue can be closed until it actually solved in a distributed pytorch...

**Comment by sitsofe** [comment]
@mingfeima sorry to nag but it's there any follow-up on this one after my previous comment?

**Comment by mingfeima** [comment]
@sitsofe yes, it is being tracked inside intel. Hopefully this will be fixed with next mkl-dnn upgrade.

**Comment by XiaobingSuper** [comment]
@sakaia , I verify that MKLDNN v0.21.1 can solve this problem, we have upgrade mkldnn to [0.21.1](https://github.com/pytorch/pytorch/pull/27597), you can test it your side by `pip install --pre torch torchvision -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html` or build the pytorch source code according to [README.md](https://github.com/pytorch/pytorch). Thanks!

**Comment by sitsofe** [comment]
I've just retested and this is working with the latest nightly using the steps from https://github.com/pytorch/pytorch/issues/20583#issuecomment-536278432 so I agree this has been fixed!
